engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnDecisionBehavior.java

```diff
@@ -89,7 +89,24 @@ public BpmnDecisionBehavior(
             .flatMap(drg -> parseDrg(drg.getResource()))
             // all the above failures have the same error type and the correct scope
             .mapLeft(f -> new Failure(f.getMessage(), ErrorType.CALLED_DECISION_ERROR, scopeKey))
-            .flatMap(drg -> evaluateDecisionInDrg(drg, element.getDecisionId(), scopeKey));
+            .flatMap(
+                drg -> {
+                  final var evaluationResult =
+                      evaluateDecisionInDrg(drg, element.getDecisionId(), scopeKey);
+
+                  final var decision = decisionOrFailure.get();
+                  writeDecisionEvaluationEvent(decision, evaluationResult, context);
+
+                  if (evaluationResult.isFailure()) {
+                    return Either.left(
+                        new Failure(
+                            evaluationResult.getFailureMessage(),
+                            ErrorType.CALLED_ELEMENT_ERROR,
+                            scopeKey));
+                  } else {
+                    return Either.right(evaluationResult);
+                  }
```

Comment:
:thought_balloon: The code looks nicer if the `Either` mapping is hidden inside the function. But on the other side, I don't want to hide the writing of the evaluation event because it is an important part.

Reply:
💭 Currently the writeDecisionEvaluationEvent (which is a side-effect) happens inside the flatmap. It could be an idea to extract that to an ifRight. Something like:

```java
public Either<Failure, DecisionEvaluationResult> evaluateDecision(
    final ExecutableCalledDecision element, final BpmnElementContext context) {
  final var scopeKey = context.getElementInstanceKey();

  // todo(#8571): avoid parsing drg every time
  final var decisionOrFailure = findDecisionById(element.getDecisionId());
  final var resultOrFailure =
      decisionOrFailure
          .flatMap(this::findDrgByDecision)
          .mapLeft(
              failure ->
                  new Failure(
                      "Expected to evaluate decision '%s', but %s"
                          .formatted(element.getDecisionId(), failure.getMessage())))
          .flatMap(drg -> parseDrg(drg.getResource()))
          // all the above failures have the same error type and the correct scope
          .mapLeft(f -> new Failure(f.getMessage(), ErrorType.CALLED_DECISION_ERROR, scopeKey))
          .map(drg -> evaluateDecisionInDrg(drg, element.getDecisionId(), scopeKey));

  resultOrFailure.ifRight(
      evaluationResult -> {
        final var decision = decisionOrFailure.get();
        writeDecisionEvaluationEvent(decision, evaluationResult, context);
      });

  // todo: could have a better name
  final var evaluatedResultOrFailure =
      resultOrFailure.flatMap(
          evaluationResult -> {
            if (evaluationResult.isFailure()) {
              return Either.left(
                  new Failure(
                      evaluationResult.getFailureMessage(),
                      ErrorType.CALLED_ELEMENT_ERROR,
                      scopeKey));
            } else {
              return Either.right(evaluationResult);
            }
          });

  evaluatedResultOrFailure.ifRight(
      result -> {
        // The output mapping behavior determines what to do with the decision result. Since the
        // output mapping may fail and raise an incident, we need to write the variable to a
        // record. This is because we want to evaluate the decision on element activation, while
        // the output mapping happens on element completion. We don't want to re-evaluate the
        // decision for output mapping related incidents.
        triggerProcessEventWithResultVariable(context, element.getResultVariable(), result);
      });

  return evaluatedResultOrFailure;
}
```

Reply:
It is an interesting idea. Thank you for sharing.
But I'm not sure if it make the code more or less readable :sweat_smile: